src/diff_folded/mod.rs

```diff
@@ -0,0 +1,158 @@
+///  - A sample count for the given stack.
+///
+/// The output written to the `writer` will be similar to the inputs, except there will be two
+/// sample count columns -- one for each profile.
+pub fn from_readers<R1, R2, W>(opt: &Options, reader1: R1, reader2: R2, writer: W) -> io::Result<()>
+where
+    R1: BufRead,
+    R2: BufRead,
+    W: Write,
+{
+    let mut stack_counts = HashMap::new();
+    let total1 = parse_stack_counts(&opt, &mut stack_counts, reader1, true)?;
+    let total2 = parse_stack_counts(&opt, &mut stack_counts, reader2, false)?;
+    if opt.normalize && total1 != total2 {
+        for counts in stack_counts.values_mut() {
+            counts.first = counts.first * total2 / total1;
+        }
+    }
+    write_stacks(&stack_counts, writer)
+}
+
+/// Produce an output that can be used to generate a differential flame graph from
+/// a before and an after profile.
+///
+/// See [`from_readers`] for the input and output formats.
+pub fn from_files<P1, P2, W>(
+    opt: &Options,
+    filename1: P1,
+    filename2: P2,
+    writer: W,
+) -> io::Result<()>
+where
+    P1: AsRef<Path>,
+    P2: AsRef<Path>,
+    W: Write,
+{
+    let file1 = File::open(filename1)?;
+    let reader1 = io::BufReader::with_capacity(READER_CAPACITY, file1);
+    let file2 = File::open(filename2)?;
+    let reader2 = io::BufReader::with_capacity(READER_CAPACITY, file2);
+    from_readers(opt, reader1, reader2, writer)
+}
+
+// Populate stack_counts based on lines from the reader and returns the sum of the sample counts.
+fn parse_stack_counts<R>(
+    opt: &Options,
+    stack_counts: &mut HashMap<String, Counts>,
+    mut reader: R,
+    is_first: bool,
+) -> io::Result<usize>
+where
+    R: BufRead,
+{
+    let mut total = 0;
+    let mut line = String::new();
+    loop {
+        line.clear();
+
+        if reader.read_line(&mut line)? == 0 {
+            break;
+        }
+
+        if let Some((stack, count)) = parse_line(&line, opt.strip_hex) {
+            let mut counts = stack_counts.entry(stack).or_default();
+            if is_first {
+                counts.first += count;
+            } else {
+                counts.second += count;
+            }
+            total += count;
+        } else {
+            warn!("Unable to parse line: {}", line);
+        }
+    }
+
+    Ok(total)
+}
+
+// Write three-column lines with the folded stack trace and two value columns,
+// one for each profile.
+fn write_stacks<W>(stack_counts: &HashMap<String, Counts>, mut writer: W) -> io::Result<()>
+where
+    W: Write,
+{
+    for (stack, &Counts { first, second }) in stack_counts {
+        writeln!(writer, "{} {} {}", stack, first, second)?;
+    }
+    Ok(())
+}
+
+// Parse stack and sample count from line.
+fn parse_line(line: &str, strip_hex: bool) -> Option<(String, usize)> {
+    let counti = line.rfind(' ')?;
+    let count = &line[(counti + 1)..].trim_end();
+    let count = count.parse::<usize>().ok()?;
```

Comment:
Won't this choke on fractional samples?

Reply:
Yup, sure will.

Reply:
I changed the sample count type to `f64` so it can handle fractional samples, but now the formatting isn't matching the Perl version.

Rust: `14.234398324872284`
Perl: `14.2343983248723`

Is there a built-in way in Rust to format with a maximum number of decimal places, but with no trailing zeroes?

Reply:
Ah, sorry, I was thinking we should ditch the fractional part [like we do in `flamegraph`](https://github.com/jonhoo/inferno/blob/643cf821ee271e968259fafd1499130186499c59/src/flamegraph/merge.rs#L191L196).

Reply:
OK, I guess if we're already different in flamegraph, it doesn't hurt here either.